Tools/TestWebKitAPI/Tests/WebCore/MixedContentChecker.cpp

```diff
@@ -0,0 +1,104 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+
+#include <WebCore/MixedContentChecker.h>
+
+using namespace WebCore;
+
+namespace TestWebKitAPI {
+
+TEST(MixedContentChecker, CanModifyRequest)
+{
+    URL url { "http://example.com/cat.jpg"_s };
+    FetchOptions::Destination destination = FetchOptions::Destination::Image;
+    Initiator initiator = Initiator::EmptyString;
+    bool shouldUpgradeIPAddressAndLocalhostForTesting = false;
+
+    ASSERT_TRUE(MixedContentChecker::canModifyRequest(
+        url,
+        destination,
+        initiator,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
+
+    // 4.1.1 request’s URL is a potentially trustworthy URL.
+    ASSERT_FALSE(MixedContentChecker::canModifyRequest(
+        URL { "https://example.com/cat.jpg"_s },
+        destination,
+        initiator,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
+
+    // 4.1.2 request’s URL’s host is an IP address.
+    URL ipAddressURL { "http://192.0.1.36"_s };
+
+    ASSERT_FALSE(MixedContentChecker::canModifyRequest(
+        ipAddressURL,
+        destination,
+        initiator,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
+
+    ASSERT_TRUE(MixedContentChecker::canModifyRequest(
+        ipAddressURL,
+        destination,
+        initiator,
+        true
+    ));
+
+    // 4.1.4 request’s destination is not "image", "audio", or "video".
+    ASSERT_TRUE(MixedContentChecker::canModifyRequest(
+        url,
+        FetchOptions::Destination::Audio,
+        initiator,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
+
+    ASSERT_TRUE(MixedContentChecker::canModifyRequest(
+        url,
+        FetchOptions::Destination::Video,
+        initiator,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
+
+    ASSERT_FALSE(MixedContentChecker::canModifyRequest(
+        url,
+        FetchOptions::Destination::Font,
+        initiator,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
+
+    // 4.1.5 request’s destination is "image" and request’s initiator is "imageset".
+    ASSERT_FALSE(MixedContentChecker::canModifyRequest(
+        url,
+        destination,
+        Initiator::Imageset,
+        shouldUpgradeIPAddressAndLocalhostForTesting
+    ));
```

Comment:
Can you two more tests that use a custom scheme? One test where the scheme is not registered, and a second test where the scheme is registered?

Reply:
Yeah! Sorry, that's the important test for the change I've made.